pydantic/_internal/_typing_extra.py

```diff
@@ -203,24 +203,25 @@ def parent_frame_namespace(*, parent_depth: int = 2, force: bool = False) -> dic
     return frame.f_locals
 
 
-def add_module_globals(obj: Any, globalns: dict[str, Any] | None = None) -> dict[str, Any]:
+def get_module_ns_of(obj: Any) -> dict[str, Any]:
+    """Get the namespace of the module where the object is defined.
+
+    Caution: this function does not return a copy of the module namespace, so it should not be mutated.
+    The burden of enforcing this is on the caller.
+    """
     module_name = getattr(obj, '__module__', None)
     if module_name:
         try:
-            module_globalns = sys.modules[module_name].__dict__
+            return sys.modules[module_name].__dict__
         except KeyError:
             # happens occasionally, see https://github.com/pydantic/pydantic/issues/2363
-            ns = {}
-        else:
-            ns = {**module_globalns, **globalns} if globalns else module_globalns.copy()
-    else:
-        ns = globalns or {}
-
-    return ns
+            return {}
+    return {}
 
 
-def get_cls_types_namespace(cls: type[Any], parent_namespace: dict[str, Any] | None = None) -> dict[str, Any]:
-    ns = add_module_globals(cls, parent_namespace)
+def merge_cls_and_parent_ns(cls: type[Any], parent_namespace: dict[str, Any] | None = None) -> dict[str, Any]:
+    ns = get_module_ns_of(cls).copy()
+    ns.update(parent_namespace or {})
```

Comment:
I would do

```suggestion
    if parent_namespace is not None:
        ns.update(parent_namespace)
```

More stylistic than anything else, although this improves performance by a factor of 10 when `parent_namespace` is `None` (probably not relevant, but still good to have :) ).